packages/strategy-tests/src/lib.rs

```diff
@@ -569,19 +569,36 @@ impl Strategy {
                         document_type,
                         contract,
                     }) => {
-                        // Get the first 10 identities who are eligible to submit documents for this contract
-                        let first_10_eligible_identities: Vec<Identity> = current_identities
+                        // Filter and collect eligible identities
+                        // Eligible identities have less than 24 documents in the mempool for this contract
+                        let eligible_identities: Vec<Identity> = current_identities
                             .iter()
                             .filter(|identity| {
                                 mempool_document_counter
                                     .get(&(identity.id(), contract.id()))
                                     .unwrap_or(&0)
                                     < &24u64
                             })
-                            .take(10)
                             .cloned()
                             .collect();
 
+                        let k = 10; // Number of random elements to select
+                        let mut reservoir = Vec::new();
+
+                        // Reservoir sampling algorithm
+                        for (i, identity) in eligible_identities.into_iter().enumerate() {
+                            if i < k {
+                                reservoir.push(identity);
+                            } else {
+                                let j = rng.gen_range(0..=i);
```

Comment:
That's very inefficient in case the current identities slice is big. Collect is a pretty heavy operation, and calling the randomizer in a loop for each element is a bad idea.

For example, you can use https://docs.rs/rand/latest/rand/seq/trait.IteratorRandom.html

Reply:
Fixed